pkg/scheduler/controller/shoot/reconciler.go

```diff
@@ -208,6 +215,34 @@ func filterSeedsMatchingProviders(cloudProfile *gardencorev1beta1.CloudProfile,
 	return matchingSeeds, nil
 }
 
+// filterMultiZonalSeeds filters seeds based on label, if shoot requires a multi-zonal control plane
+func filterMultiZonalSeeds(seedList []gardencorev1beta1.Seed, shoot *gardencorev1beta1.Shoot) ([]gardencorev1beta1.Seed, error) {
+	// TODO:(shreyas-s-rao): switch from shoot annotation to shoot spec field for beta release of HAControlPlanes
+	var multiZonalSeeds []gardencorev1beta1.Seed
+	var nonMultiZonalSeeds []gardencorev1beta1.Seed
+
+	for _, seed := range seedList {
+		if metav1.HasLabel(seed.ObjectMeta, v1beta1constants.LabelSeedMultiZonal) {
+			multiZonalSeeds = append(multiZonalSeeds, seed)
+		} else {
+			nonMultiZonalSeeds = append(nonMultiZonalSeeds, seed)
+		}
+	}
+
+	for _, knownFeature := range gardenerschedulerfeatures.FeatureGate.KnownFeatures() {
+		if knownFeature == string(features.HAControlPlanes) {
+			if gardenerschedulerfeatures.FeatureGate.Enabled(features.HAControlPlanes) && metav1.HasAnnotation(shoot.ObjectMeta, v1beta1constants.ShootAlphaControlPlaneHighAvailability) && shoot.ObjectMeta.Annotations[v1beta1constants.ShootAlphaControlPlaneHighAvailability] == v1beta1constants.ShootAlphaControlPlaneHighAvailabilityMultiZonal {
```

Comment:
what happens if the HAControlPlanes feature is enabled and annotation is set but its value is not `multi-zonal`, in that case would you only return non-multizonal seeds?

Reply:
Yes. Only if the value is `multi-zonal`, it will return multi-zonal seeds. If not, that means that the etcd still needs to be HA, but within a single zone, and any non-multi-zonal seed would suffice for such shoots.
The hard filtering here is to ensure that `non-HA` and `HA non-MAZ` shoots don't take up all the quota of a `MAZ` seed, because we want such seeds to _only_ host `HA MAZ` shoots.